### PR TITLE
Revalidate Type

### PR DESCRIPTION
**Pull request by wyattjoh**
This adds a new `Revalidate` type which is used internally by Next.js to associate the user inputted value of `revalidate` from `getStaticProps` or the exported `revalidate` variable in app directory.

### CODE REVIEW

**Comment by wyattjoh** [comment]
Current dependencies on/for this PR:
* canary
  * **PR #56763** <a href="https://app.graphite.dev/github/pr/vercel/next.js/56763" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/56763?utm_source=stack-comment).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 154992fce28f8b64c6060d48d67fc461e47aabed

`pnpm test-dev test/development/acceptance-app/ReactRefreshRequire.test.ts` 
- ReactRefreshRequire app > re-runs accepted modules
- ReactRefreshRequire app > propagates a hot update to closest accepted module
- ReactRefreshRequire app > propagates hot update to all inverse dependencies

<details>
<summary>Expand output</summary>

  ● ReactRefreshRequire app › re-runs accepted modules

    expect(received).toEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 1

      Array [
        "init BarV2",
    +   "init BarV2",
      ]

      50 |       `window.log.push('init BarV2'); export default function Bar() { return null; };`
      51 |     )
    > 52 |     expect(await session.evaluate(() => (window as any).log)).toEqual([
         |                                                               ^
      53 |       'init BarV2',
      54 |     ])
      55 |

      at Object.<anonymous> (development/acceptance-app/ReactRefreshRequire.test.ts:52:63)

  ● ReactRefreshRequire app › propagates a hot update to closest accepted module

    expect(received).toEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 2

      Array [
        "init FooV1",
        "init BarV2",
    +   "init FooV1",
    +   "init BarV2",
      ]

      109 |     await session.evaluate(() => ((window as any).log = []))
      110 |     await session.patch('./bar.js', `window.log.push('init BarV2');`)
    > 111 |     expect(await session.evaluate(() => (window as any).log)).toEqual([
          |                                                               ^
      112 |       // // FIXME: Metro order:
      113 |       // 'init BarV2',
      114 |       // 'init FooV1',

      at Object.<anonymous> (development/acceptance-app/ReactRefreshRequire.test.ts:111:63)

  ● ReactRefreshRequire app › propagates hot update to all inverse dependencies

    expect(received).toEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 3

      Array [
        "init LeafV2",
        "init MiddleAV1",
        "init MiddleBV1",
    +   "init LeafV2",
    +   "init MiddleAV1",
    +   "init MiddleBV1",
      ]

      264 |       `log.push('init LeafV2'); export default {};`
      265 |     )
    > 266 |     expect(await session.evaluate(() => (window as any).log)).toEqual([
          |                                                               ^
      267 |       'init LeafV2',
      268 |       'init MiddleAV1',
      269 |       'init MiddleBV1',

      at Object.<anonymous> (development/acceptance-app/ReactRefreshRequire.test.ts:266:63)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js refactor/revalidate-type | Change |
| - | - | - | - |
| buildDuration | 10.1s | 10.3s | ⚠️ +132ms |
| buildDurationCached | 6.1s | 6.3s | ⚠️ +115ms |
| nodeModulesSize | 172 MB | 172 MB | ⚠️ +2.56 kB |
| nextStartRea..uration (ms) | 902ms | 894ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/revalidate-type | Change |
| - | - | - | - |
| 199-HASH.js gzip | 27.5 kB | 27.5 kB | N/A |
| 3f784ff6-HASH.js gzip | 50.9 kB | 50.9 kB | ✓ |
| 99.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.3 kB | 45.3 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 252 B | N/A |
| main-HASH.js gzip | 32.9 kB | 32.9 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 96.4 kB | 96.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/revalidate-type | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/revalidate-type | Change |
| - | - | - | - |
| _app-HASH.js gzip | 206 B | 205 B | N/A |
| _error-HASH.js gzip | 182 B | 180 B | N/A |
| amp-HASH.js gzip | 506 B | 505 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | N/A |
| edge-ssr-HASH.js gzip | 260 B | 259 B | N/A |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.35 kB | 4.35 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.63 kB | N/A |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 384 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.08 kB | 1.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/revalidate-type | Change |
| - | - | - | - |
| _buildManifest.js gzip | 485 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/revalidate-type | Change |
| - | - | - | - |
| index.html gzip | 528 B | 530 B | N/A |
| link.html gzip | 540 B | 543 B | N/A |
| withRouter.html gzip | 524 B | 526 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/revalidate-type | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.7 kB | 93.7 kB | N/A |
| page.js gzip | 152 kB | 152 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/revalidate-type | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 622 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 22.9 kB | 22.9 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 154992fce28f8b64c6060d48d67fc461e47aabed

File: packages/next/src/server/render.tsx [hunk 1]
```diff
@@ -951,24 +954,28 @@ export async function renderToHTMLImpl(
...
               `\nTo revalidate as soon as possible, you can set the value to \`1\`.`
           )
-        } else if (data.revalidate > 31536000) {
-          // if it's greater than a year for some reason error
-          console.warn(
-            `Warning: A page's revalidate option was set to more than a year for ${req.url}. This may have been done in error.` +
-              `\nTo only run getStaticProps at build-time and not revalidate at runtime, you can set \`revalidate\` to \`false\`!`
-          )
+        } else {
+          if (data.revalidate > 31536000) {
```
Comment (styfle) [inline packages/next/src/server/render.tsx:958]: Was this changed because TS didn't like it or something?

**Review by styfle — APPROVED** [review]

**Pull request merged**